File: android-katha-ai/app/src/main/java/com/rork/kathaai/model/AnalyticsModels.kt
package com.rork.kathaai.model

import kotlinx.serialization.Serializable

// MARK: - Time Range

enum class AnalyticsTimeRange(val label: String, val dayCount: Int?) {
    SEVEN_DAYS("7 days", 7),
    THIRTY_DAYS("30 days", 30),
    ALL_TIME("All time", null);
}

// MARK: - Daily Reads

@Serializable
data class DailyReads(
    val dateISO: String,
    val reads: Int,
    val uniqueReads: Int
)

// MARK: - Chapter Reads

@Serializable
data class ChapterReads(
    val chapterId: String,
    val reads: Int,
    val uniqueReads: Int,
    val completionRate: Double
)

// MARK: - Story Analytics

@Serializable
data class StoryAnalytics(
    val dailyReads: List<DailyReads> = emptyList(),
    val chapterReads: List<ChapterReads> = emptyList(),
    val milestonesCrossed: Set<String> = emptySet(),
    val creditsEarnedFromReads: Int = 0,
    val followersGainedFromStory: Int = 0,
    val authorFollowersGainedViaStory: Int = 0,
    val lastUpdated: Long = System.currentTimeMillis()
)

// MARK: - Milestone Key

enum class MilestoneKey(val key: String, val label: String, val icon: String) {
    PUBLISH("publish", "Published", "paperplane"),
    FIRST_READ("first_read", "First read ✨", "eye"),
    READS_10("reads_10", "10 reads", "book"),
    READS_100("reads_100", "100 reads", "book"),
    READS_500("reads_500", "500 reads", "book"),
    READS_1000("reads_1000", "1,000 reads", "book"),
    READS_10000("reads_10000", "10,000 reads", "book"),
    FOLLOWERS_10("followers_10", "10 story followers", "heart"),
    FOLLOWERS_100("followers_100", "100 story followers", "heart"),
    FOLLOWERS_500("followers_500", "500 story followers", "heart"),
    FOLLOWERS_1000("followers_1000", "1,000 story followers", "heart"),
    FIRST_CREDIT("first_credit", "First credit earned ✨", "credits"),
    CREDITS_10("credits_10", "10 credits from reads", "credits"),
    CREDITS_50("credits_50", "50 credits from reads", "credits"),
    CREDITS_100("credits_100", "100 credits from reads", "credits"),
    AUTHOR_FOLLOWERS_100("author_followers_100", "100 people follow you", "person_2");

    companion object {
        fun fromKey(key: String): MilestoneKey? = entries.firstOrNull { it.key == key }
    }

    fun isCrossed(totalReads: Int, storyFollowers: Int, creditsEarned: Int, isPublished: Boolean, authorFollowers: Int): Boolean {
        return when (this) {
            PUBLISH -> isPublished
            FIRST_READ -> totalReads >= 1
            READS_10 -> totalReads >= 10
            READS_100 -> totalReads >= 100
            READS_500 -> totalReads >= 500
            READS_1000 -> totalReads >= 1000
            READS_10000 -> totalReads >= 10000
            FOLLOWERS_10 -> storyFollowers >= 10
            FOLLOWERS_100 -> storyFollowers >= 100
            FOLLOWERS_500 -> storyFollowers >= 500
            FOLLOWERS_1000 -> storyFollowers >= 1000
            FIRST_CREDIT -> creditsEarned >= 1
            CREDITS_10 -> creditsEarned >= 10
            CREDITS_50 -> creditsEarned >= 50
            CREDITS_100 -> creditsEarned >= 100
            AUTHOR_FOLLOWERS_100 -> authorFollowers >= 100
        }
    }

    fun celebrationTitle(): String = when (this) {
        PUBLISH -> "Your story is live ✨"
        FIRST_READ -> "Your first reader found you"
        READS_10 -> "10 reads on your story 📖"
        READS_100 -> "100 reads on your story 🎉"
        READS_500 -> "500 reads on your story 🔥"
        READS_1000 -> "1,000 reads on your story 💫"
        READS_10000 -> "10,000 reads. Katha royalty. 👑"
        FOLLOWERS_10 -> "10 followers on your story"
        FOLLOWERS_100 -> "100 followers ✨"
        FOLLOWERS_500 -> "500 followers 🌟"
        FOLLOWERS_1000 -> "1,000 followers 👑"
        FIRST_CREDIT -> "First credit earned from a reader ✨"
        CREDITS_10 -> "10 credits earned 💰"
        CREDITS_50 -> "50 credits earned 🌟"
        CREDITS_100 -> "100 credits earned 👑"
        AUTHOR_FOLLOWERS_100 -> "100 people follow you"
    }

    fun celebrationSubtitle(storyTitle: String): String = when (this) {
        PUBLISH -> "'$storyTitle' just went out into the world"
        FIRST_READ -> "Someone just read '$storyTitle'"
        READS_10 -> "You're being read."
        READS_100 -> "That's a hundred people spent time with your words."
        READS_500 -> "You've officially got an audience."
        READS_1000 -> "A milestone worth celebrating. Keep writing."
        READS_10000 -> "'$storyTitle' has hit a rare milestone."
        FOLLOWERS_10 -> "Ten people are waiting for your next chapter."
        FOLLOWERS_100 -> "A hundred readers are following '$storyTitle'."
        FOLLOWERS_500 -> "Impressive. '$storyTitle' has a real following."
        FOLLOWERS_1000 -> "'$storyTitle' is one of Katha's most-followed."
        FIRST_CREDIT -> "'$storyTitle' just paid for its next chapter."
        CREDITS_10 -> "'$storyTitle' has funded your next 10 generations."
        CREDITS_50 -> "'$storyTitle' is a real earner. Keep writing."
        CREDITS_100 -> "'$storyTitle' is bankrolling your creative practice."
        AUTHOR_FOLLOWERS_100 -> "You've built a real audience on Katha."
    }
}

// MARK: - Unseen Milestone

@Serializable
data class UnseenMilestone(
    val id: String,
    val milestoneKey: String,
    val storyId: String,
    val storyTitle: String,
    val timestamp: Long = System.currentTimeMillis()
)

// MARK: - Stat Trend

data class StatTrend(val delta: Int) {
    val isPositive: Boolean get() = delta > 0
    val label: String get() = if (delta == 0) "No change" else "${if (delta > 0) "+" else ""}$delta vs previous period"
}

// MARK: - Sort Metric

enum class MilestoneSortMetric(val label: String) {
    READS("By reads"),
    CREDITS("By credits"),
    FOLLOWERS("By followers")
}

File: ios-katha-ai-create-stories/KathaAICreateStories/Models/AnalyticsModels.swift
//
//  AnalyticsModels.swift
//  KathaAICreateStories
//

import Foundation

// MARK: - Time Range

enum AnalyticsTimeRange: Int, CaseIterable, Identifiable {
    case sevenDays = 0
    case thirtyDays = 1
    case allTime = 2

    var id: Int { rawValue }

    var label: String {
        switch self {
        case .sevenDays: "7 days"
        case .thirtyDays: "30 days"
        case .allTime: "All time"
        }
    }

    var dayCount: Int? {
        switch self {
        case .sevenDays: 7
        case .thirtyDays: 30
        case .allTime: nil
        }
    }
}

// MARK: - Daily Reads

struct DailyReads: Identifiable, Codable, Hashable {
    var id: String { dateISO }
    let dateISO: String
    let reads: Int
    let uniqueReads: Int

    var date: Date {
        AnalyticsDateFormatter.shared.date(from: dateISO) ?? Date()
    }
}

// MARK: - Chapter Reads

struct ChapterReads: Codable, Hashable {
    let chapterId: String
    let reads: Int
    let uniqueReads: Int
    let completionRate: Double
}

// MARK: - Story Analytics

struct StoryAnalytics: Codable, Hashable {
    var dailyReads: [DailyReads]
    var chapterReads: [ChapterReads]
    var milestonesCrossed: Set<String>
    var creditsEarnedFromReads: Int
    var followersGainedFromStory: Int
    var authorFollowersGainedViaStory: Int
    var lastUpdated: Date
}

// MARK: - Milestone Key

enum MilestoneKey: String, CaseIterable, Identifiable {
    case publish
    case firstRead
    case reads10
    case reads100
    case reads500
    case reads1000
    case reads10000
    case followers10
    case followers100
    case followers500
    case followers1000
    case firstCredit
    case credits10
    case credits50
    case credits100
    case authorFollowers100

    var id: String { rawValue }

    var label: String {
        switch self {
        case .publish: "Published"
        case .firstRead: "First read ✨"
        case .reads10: "10 reads"
        case .reads100: "100 reads"
        case .reads500: "500 reads"
        case .reads1000: "1,000 reads"
        case .reads10000: "10,000 reads"
        case .followers10: "10 story followers"
        case .followers100: "100 story followers"
        case .followers500: "500 story followers"
        case .followers1000: "1,000 story followers"
        case .firstCredit: "First credit earned ✨"
        case .credits10: "10 credits from reads"
        case .credits50: "50 credits from reads"
        case .credits100: "100 credits from reads"
        case .authorFollowers100: "100 people follow you"
        }
    }

    var icon: String {
        switch self {
        case .publish: "paperplane"
        case .firstRead: "eye"
        case .reads10: "book"
        case .reads100: "book"
        case .reads500: "book"
        case .reads1000: "book"
        case .reads10000: "book"
        case .followers10: "heart"
        case .followers100: "heart"
        case .followers500: "heart"
        case .followers1000: "heart"
        case .firstCredit: "credits"
        case .credits10: "credits"
        case .credits50: "credits"
        case .credits100: "credits"
        case .authorFollowers100: "person.2"
        }
    }

    /// Celebration banner title copy
    var celebrationTitle: String {
        switch self {
        case .publish: "Your story is live ✨"
        case .firstRead: "Your first reader found you"
        case .reads10: "10 reads on your story 📖"
        case .reads100: "100 reads on your story 🎉"
        case .reads500: "500 reads on your story 🔥"
        case .reads1000: "1,000 reads on your story 💫"
        case .reads10000: "10,000 reads. Katha royalty. 👑"
        case .followers10: "10 followers on your story"
        case .followers100: "100 followers ✨"
        case .followers500: "500 followers 🌟"
        case .followers1000: "1,000 followers 👑"
        case .firstCredit: "First credit earned from a reader ✨"
        case .credits10: "10 credits earned 💰"
        case .credits50: "50 credits earned 🌟"
        case .credits100: "100 credits earned 👑"
        case .authorFollowers100: "100 people follow you"
        }
    }

    /// Celebration banner subtitle copy (with story title placeholder)
    func celebrationSubtitle(storyTitle: String) -> String {
        switch self {
        case .publish: "'\(storyTitle)' just went out into the world"
        case .firstRead: "Someone just read '\(storyTitle)'"
        case .reads10: "You're being read."
        case .reads100: "That's a hundred people spent time with your words."
        case .reads500: "You've officially got an audience."
        case .reads1000: "A milestone worth celebrating. Keep writing."
        case .reads10000: "'\(storyTitle)' has hit a rare milestone."
        case .followers10: "Ten people are waiting for your next chapter."
        case .followers100: "A hundred readers are following '\(storyTitle)'."
        case .followers500: "Impressive. '\(storyTitle)' has a real following."
        case .followers1000: "'\(storyTitle)' is one of Katha's most-followed."
        case .firstCredit: "'\(storyTitle)' just paid for its next chapter."
        case .credits10: "'\(storyTitle)' has funded your next 10 generations."
        case .credits50: "'\(storyTitle)' is a real earner. Keep writing."
        case .credits100: "'\(storyTitle)' is bankrolling your creative practice."
        case .authorFollowers100: "You've built a real audience on Katha."
        }
    }

    /// Check if this milestone is crossed given the metrics
    func isCrossed(totalReads: Int, storyFollowers: Int, creditsEarned: Int, isPublished: Bool, authorFollowers: Int) -> Bool {
        switch self {
        case .publish: isPublished
        case .firstRead: totalReads >= 1
        case .reads10: totalReads >= 10
        case .reads100: totalReads >= 100
        case .reads500: totalReads >= 500
        case .reads1000: totalReads >= 1000
        case .reads10000: totalReads >= 10000
        case .followers10: storyFollowers >= 10
        case .followers100: storyFollowers >= 100
        case .followers500: storyFollowers >= 500
        case .followers1000: storyFollowers >= 1000
        case .firstCredit: creditsEarned >= 1
        case .credits10: creditsEarned >= 10
        case .credits50: creditsEarned >= 50
        case .credits100: creditsEarned >= 100
        case .authorFollowers100: authorFollowers >= 100
        }
    }
}

// MARK: - Unseen Milestone (queue for celebration banner)

struct UnseenMilestone: Identifiable, Codable, Hashable {
    let id: String
    let milestoneKey: String
    let storyId: String
    let storyTitle: String
    let timestamp: Date
}

// MARK: - Stat Trend

struct StatTrend: Hashable {
    let delta: Int
    let isPositive: Bool

    var label: String {
        if delta == 0 { return "No change" }
        let prefix = delta > 0 ? "+" : ""
        return "\(prefix)\(delta) vs previous period"
    }

    var color: String {
        if delta > 0 { return "success" }
        if delta < 0 { return "error" }
        return "secondary"
    }
}

// MARK: - Analytics Date Formatter

enum AnalyticsDateFormatter {
    static let shared: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withFullDate]
        return f
    }()

    static let displayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "MMM d"
        return f
    }()

    static let publishFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateStyle = .medium
        f.timeStyle = .none
        return f
    }()
}
